mrmustard/lab_dev/bb.py

```diff
@@ -0,0 +1,414 @@
+
+        Arguments:
+            other: Component
+
+        Returns:
+            int: contraction cost in approx FLOPS
+
+        TODO: this will need to be updated once we have the poly x exp ansatz.
+        TODO: be more precise on costs (profile properly? use wall time?)
+        """
+        idxA, idxB = self.wires.contracted_indices(other.wires)
+        m = len(idxA)  # same as len(idxB)
+        nA, nB = len(self.shape) - m, len(other.shape) - m
+
+        if self.repr == "Bargmann" and other.repr == "Bargmann":
+            cost = (  # +1s to include vector part)
+                m * m * m  # M inverse
+                + (m + 1) * m * nA  # left matmul
+                + (m + 1) * m * nB  # right matmul
+                + (m + 1) * m  # addition
+                + m * m * m  # determinant of M
+            )
+        else:  # otherwise need to use fock cost
+            prod_A = np.prod([s for i, s in enumerate(self.shape) if i not in idxA])
+            prod_B = np.prod([s for i, s in enumerate(other.shape) if i not in idxB])
+            prod_contracted = np.prod(
+                [min(self.shape[i], other.shape[j]) for i, j in zip(idxA, idxB)]
+            )
+            cost = (
+                prod_A * prod_B * prod_contracted  # matmul
+                + np.prod(self.shape) * (self.repr == "Bargmann")  # conversion
+                + np.prod(other.shape) * (other.repr == "Bargmann")  # conversion
+            )
+        return int(cost)
+
+    def __matmul__(self, other) -> Component:
+        """Returns the contracted Component."""
+        new_wires, perm = self.wires @ other.wires
+        idxA, idxB = self.wires.contracted_indices(other.wires)
+        shape_A = [n for i, n in enumerate(self.shape) if i not in idxA]
+        shape_B = [n for i, n in enumerate(other.shape) if i not in idxB]
+        shape = shape_A + shape_B
+        new_shape = [shape[p] for p in perm]
+        new_component = Component(
+            "Bargmann" if self.repr == other.repr == "Bargmann" else "Fock",
+            new_wires,
+            new_shape,
+            f"({self.name}@{other.name})",
+            self.contraction_cost(other) + self.cost + other.cost,
+        )
+        return new_component
+
+    def __repr__(self):
+        return f"{self.name}({self.shape}, {self.wires})"
+
+
+class Graph(nx.DiGraph):
+    """Power pack for nx.DiGraph with additional attributes and methods."""
+
+    def __init__(self, solution: tuple[Edge, ...] = (), costs: tuple[int, ...] = ()):
+        super().__init__()
+        self.solution = solution
+        self.costs = costs
+
+    @property
+    def cost(self) -> int:
+        return sum(self.costs)
+
+    def __lt__(self, other: Graph) -> bool:
+        return self.cost < other.cost
+
+    def __hash__(self) -> int:
+        return hash(
+            tuple(self.nodes)
+            + tuple(self.edges)
+            + tuple(self.solution)  # do we want this?
```

Comment:
It depends on what we use this equality for. Probably since we use it mainly to prune the priority queue, it's best if we consider them to be different.

Reply:
perfect! I think removing the comment here would be good then